### PR TITLE
RD-3763 - sqlalchemy

### PR DESCRIPTION
**Pull request by saartochner**

### CODE REVIEW

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            "connectionParameters": {
+                "host": conn.engine.url.host or conn.engine.url.database,
+                "port": conn.engine.url.port,
+                "database": conn.engine.url.database,
+                "user": conn.engine.url.username,
+            },
+            "query": lumigo_dumps(statement),
+            "values": lumigo_dumps(parameters),
+        }
+    )
```
Comment (nirLumigo) [inline src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:37]: Shouldn't it be with `safe_execute`?

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def execute_wrapper(func, instance, args, kwargs):
+    result = func(*args, **kwargs)
+    with lumigo_safe_execute("sqlalchemy: listen to engine"):
+        listen(result, "before_cursor_execute", _before_cursor_execute)
+        listen(result, "after_cursor_execute", _after_cursor_execute)
+        listen(result, "handle_error", _handle_error)
```
Comment (nirLumigo) [inline src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:71]: Don't you need to do `listen` before you run `func`?

Comment (nirLumigo) [inline src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:71]: `listen` might be `None`

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    if not span:
+        get_logger().warning("Redis span ended without a record on its start")
+        return
+    span.update(
+        {
+            "ended": get_current_ms_time(),
+            "error": lumigo_dumps(
+                {
+                    "type": context.original_exception.__class__.__name__,
+                    "args": context.original_exception.args,
```
Comment (nirLumigo) [inline src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:59]: Are "name", and "args" enough? Don't we have more data about the error?

File: src/test/unit/wrappers/sql/test_sqlalchemt_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import pytest
+from lumigo_tracer.lumigo_utils import DEFAULT_MAX_ENTRY_SIZE
+from sqlalchemy.exc import OperationalError
+
+from lumigo_tracer.spans_container import SpansContainer
+from lumigo_tracer.tracer import lumigo_tracer
+from sqlalchemy import create_engine, Table, Column, Integer, String, MetaData
+from sqlalchemy.sql import select
```
Comment (nirLumigo) [inline src/test/unit/wrappers/sql/test_sqlalchemt_wrapper.py:8]: Order imports, sqlalchemy with sqlalchemy and lumigo_tracer with lumigo_tracer

File: src/test/unit/wrappers/sql/test_sqlalchemt_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import pytest
```
Comment (nirLumigo) [inline src/test/unit/wrappers/sql/test_sqlalchemt_wrapper.py:1]: Typo in the name of the file, added `t` at the end of `alchemy`

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Good job :)

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py (hunk 3 above)
Comment (saartochner) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:59]: unfortunately no,
This is the object that they give us (this is not an exception object!)

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py (hunk 1 above)
Comment (saartochner) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:37]: damn, you're right!
Good catch!!

Comment (saartochner) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:37]: I'll add test :)

File: src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py (hunk 2 above)
Comment (saartochner) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:71]: I don't run here nothing.
The lumigo tracer's setup calls to `wrappers.wrap()`, which calls to this wrapper.
I.e. no user's code ran until now :)

Comment (nirLumigo) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:74]: I think you forgot to commit this change 😅

Comment (saartochner) [reply src/lumigo_tracer/wrappers/sql/sqlalchemy_wrapper.py:74]: I did! checkout line 80

**Review by nirLumigo — APPROVED** [review]

**Pull request merged**